content/blog/2022/2022-08-02-performance-with-completablefutures.md

```diff
@@ -0,0 +1,249 @@
+version 8 of Java. CompletableFuture implements both Future
+and _[CompletionStage](https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/CompletionStage.html)_ interfaces.
+Using this class we can now create a pipeline of asynchronous operations. Let's see a simple example of how to do it :
+
+```java
+class Demo {
+    public static void main(String[] args) {
+
+        CompletableFuture.supplyAsync(() -> "hello completable future")
+                .thenApply(String::toUpperCase)
+                .thenAccept(System.out::println);
+    }
+
+}
+```
+
+In the example above we can notice how simple is to create such a pipeline. We are firstly calling the `supplyAsync`
+method which takes a `Supplier` and returns a new CompletableFuture, we are then transforming the result to uppercase
+String by calling `thenApply` method and in the end, we just print the value on the console using `thenAccept` that
+takes a Consumer as the argument. If we think for a moment, working with CompletableFuture is very similar to Java
+Streams.
+
+## Calling remote APIs without using CompletableFuture (Synchronously)
+
+In this section we will build a simple application which given a list of bank transactions with call an external service
+to categorize the transactions in base of their description. We will simulate the call of the external service by using
+a method that adds some delay before returning the category of the transaction. In the next sections we will
+incrementally change the implementation of our client application to improve the performance by using CompletableFuture.
+Let's start implementing our categorization service that declares a method called `categorizeTransaction` :
+
+```java
+public class CategorizationService {
+
+    public static Category categorizeTransaction(Transaction transaction) {
+        delay();
+        return new Category("Category_" + transaction.getId());
+    }
+
+    public static void delay() {
+        try {
+            Thread.sleep(1000L);
+        } catch (InterruptedException e) {
+            throw new RuntimeException(e);
+        }
+    }
+}
+
+public class Category {
+    private final String category;
+
+    public Category(String category) {
+        this.category = category;
+    }
+
+    @Override
+    public String toString() {
+        return "Category{" +
+                "category='" + category + '\'' +
+                '}';
+    }
+}
+
+public class Transaction {
+    private String id;
+    private String description;
+
+    public Transaction(String id, String description) {
+        this.id = id;
+        this.description = description;
+    }
+
+    public String getId() {
+        return id;
+    }
+
+    public void setId(String id) {
+        this.id = id;
+    }
+
+    public String getDescription() {
+        return description;
+    }
+
+    public void setDescription(String description) {
+        this.description = description;
+    }
+}
```

Comment:
Please add a paragraph below this code example that explains the code a bit. Not every detail, but in broad strokes.

Reply:
done